### PR TITLE
Pass _caretPrototype to prevent cache miss

### LINKED ISSUE #24522: Caret performance is poor, high GPU time per frame.
When writing to TextFields in my Flutter application, I am seeing serious drops in performance while typing. The frame rate goes from 60 -> 25 fps (and red bars). 

Oddly enough, if using a single-line textfield and typing "outside" the screen, the frame rate goes back up to ~60 (with no red bars in perf monitor). (Could also be related to number of characters, but that seems less likely given my observations).

From what I can see, the surrounding widgets does not seem to have any substantial impact, i.e. it happens on both simpler and more complex screens.

The sluggish performance is even more obvious when using the label decorations that comes with material TextFields.
However, I am seeing the same on both Cupertino and Material textfields. I have also tried with very "clean" textfields, i.e. without any text controllers or similar attached to it.

Is this expected? Or could there be any issues with the TextField widget?
(I do not see any behavior like this in native or even web applications I've built.)

## Steps to Reproduce

1. Create a `StatefulWidget` and add a `TextField` to it. In my case, the hierarchy is ~3-4 levels deep before appending the `TextField`. 
2. Start typing in the `TextField` (I've tested both on iOS 8 device and on iPhone 7 simulator).
3. To see differences in performance; make the textfield single line with a fixed width.

## Logs

```
Launching lib/main.dart on iPhone 7 in debug mode...
Starting Xcode build...
Xcode build done.                                           10,2s
https://firebase.google.com/docs/cloud-messaging/ios/client#method_swizzling_in_firebase_messaging
to ensure proper integration.
To hide this warning and ensure your app does not break, you need to add the following code to your app before calling any other Cloud Firestore methods:

let db = Firestore.firestore()
let settings = db.settings
settings.areTimestampsInSnapshotsEnabled = true
db.settings = settings

With this change, timestamps stored in Cloud Firestore will be read back as Firebase Timestamp objects instead of as system Date objects. So you will also need to update code expecting a Date to instead expect a Timestamp. For example:

// old:
let date: Date = documentSnapshot.get("created_at") as! Date
// new:
let timestamp: Timestamp = documentSnapshot.get("created_at") as! Timestamp
let date: Date = timestamp.dateValue()

Please audit all existing usages of Date when you enable the new behavior. In a future release, the behavior will be changed to the new behavior, so if you do<…>
Syncing files to device iPhone 7...
[VERBOSE-2:FlutterObservatoryPublisher.mm(145)] FlutterObservatoryPublisher is ready!

```

```
Analyzing my_app...                                         

   info • Unused import: 'package:cloud_firestore/cloud_firestore.dart' • lib/screens/account.dart:3:8 • unused_import
   info • Unused import: 'package:my_app/screens/widgets/credits.dart' • lib/screens/create_thread.dart:4:8 • unused_import
   info • This function has a return type of 'Future<String>', but doesn't end with a return statement • lib/screens/login.dart:166:3 • missing_return
   info • This class inherits from a class marked as @immutable, and therefore should be immutable (all instance fields must be final) • lib/screens/thread.dart:17:7 • must_be_immutable
   info • Avoid using braces in interpolation when not needed • lib/screens/thread.dart:178:32 • unnecessary_brace_in_string_interps


```

```
[✓] Flutter (Channel dev, v0.11.3, on Mac OS X 10.14.1 18B75, locale sv-SE)
    • Flutter version 0.11.3 at /Users/my_user/Dev/Libs/flutter
    • Framework revision 72bf075e8d (9 days ago), 2018-11-09 20:36:17 -0800
    • Engine revision 5646e86a6f
    • Dart version 2.1.0 (build 2.1.0-dev.9.3 9c07fb64c4)

[✓] Android toolchain - develop for Android devices (Android SDK 28.0.2)
    • Android SDK at /Users/my_user/Library/Android/sdk
    • Android NDK location not configured (optional; useful for native profiling support)
    • Platform android-28, build-tools 28.0.2
    • Java binary at: /Applications/Android Studio.app/Contents/jre/jdk/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)
    • All Android licenses accepted.

[✓] iOS toolchain - develop for iOS devices (Xcode 10.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • Xcode 10.1, Build version 10B61
    • ios-deploy 2.0.0
    • CocoaPods version 1.5.3

[✓] Android Studio (version 3.0)
    • Android Studio at /Applications/Android Studio.app/Contents
    ✗ Flutter plugin not installed; this adds Flutter specific functionality.
    ✗ Dart plugin not installed; this adds Dart specific functionality.
    • Java version OpenJDK Runtime Environment (build 1.8.0_152-release-915-b08)

[✓] IntelliJ IDEA Ultimate Edition (version 2018.1.6)
    • IntelliJ at /Applications/IntelliJ IDEA.app
    • Flutter plugin version 28.0.2
    • Dart plugin version 181.5616

[✓] Connected device (2 available)
    • Iphone S • <obfuscated> • ios • iOS 12.1
    • iPhone 7 • <obfuscated>     • ios • iOS 12.1 (simulator)

• No issues found!
```

### PR DESCRIPTION
**Pull request by GaryQian**
## Description

Passing Rect.zero would cache the zero caretProto, which would cause future calls to perform redundant computation.

Pass the proper _caretPrototype to ensure the cached version is not discarded.

## Related Issues

Fixes https://github.com/flutter/flutter/issues/24522

## Tests

Added perf test.

## Checklist

Before you create this PR confirm that it meets all requirements listed below by checking the relevant checkboxes (`[x]`). This will ensure a smooth and quick review process.

- [x] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [x] I signed the [CLA].
- [x] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [x] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [x] I updated/added relevant documentation (doc comments with `///`).
- [x] All existing and new tests are passing.
- [x] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [x] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

Did any tests fail when you ran them? Please read [Handling breaking changes].

- [x] No, no existing tests failed, so this is *not* a breaking change.
- [ ] Yes, this is a breaking change. *If not, delete the remainder of this section.*
   - [ ] I wrote a design doc: https://flutter.dev/go/template *Replace this with a link to your design doc's short link*
   - [ ] I got input from the developer relations team, specifically from: *Replace with the names of who gave advice*
   - [ ] I wrote a migration guide: *Replace with a link to your migration guide*

[issue database]: https://github.com/flutter/flutter/issues
[Contributor Guide]: https://github.com/flutter/flutter/wiki/Tree-hygiene#overview
[Test Coverage]: https://github.com/flutter/flutter/wiki/Test-coverage-for-package%3Aflutter
[Flutter Style Guide]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo
[Features we expect every widget to implement]: https://github.com/flutter/flutter/wiki/Style-guide-for-Flutter-repo#features-we-expect-every-widget-to-implement
[CLA]: https://cla.developers.google.com/
[Tree Hygiene]: https://github.com/flutter/flutter/wiki/Tree-hygiene
[Handling breaking changes]: https://github.com/flutter/flutter/wiki/Tree-hygiene#handling-breaking-changes

Before:
    "average_frame_build_time_millis": 3.5870540540540534,
    "90th_percentile_frame_build_time_millis": 6.709,
    "99th_percentile_frame_build_time_millis": 24.555,
    "worst_frame_build_time_millis": 24.555,
    "missed_frame_build_budget_count": 1,
    "average_frame_rasterizer_time_millis": 6.6930000000000005,
    "90th_percentile_frame_rasterizer_time_millis": 6.919,
    "99th_percentile_frame_rasterizer_time_millis": 70.695,
    "worst_frame_rasterizer_time_millis": 70.695,
    "missed_frame_rasterizer_budget_count": 2,
    "frame_count": 37,

After:
    "average_frame_build_time_millis": 3.3834864864864858,
    "90th_percentile_frame_build_time_millis": 4.756,
    "99th_percentile_frame_build_time_millis": 24.9,
    "worst_frame_build_time_millis": 24.9,
    "missed_frame_build_budget_count": 1,
    "average_frame_rasterizer_time_millis": 6.965729729729728,
    "90th_percentile_frame_rasterizer_time_millis": 8.253,
    "99th_percentile_frame_rasterizer_time_millis": 74.73,
    "worst_frame_rasterizer_time_millis": 74.73,
    "missed_frame_rasterizer_budget_count": 2,
    "frame_count": 37,

### CODE REVIEW

**Comment by GaryQian** [comment]
This change does not really result in any testable changes, as passing `Rect.zero` vs `_caretPrototype` results in the same return values. The only difference is the internally cached values hitting/missing. I don't think this is worth adding a specific benchmark for.

File: packages/flutter/lib/src/rendering/editable.dart [hunk 1]
```diff
@@ -365,7 +365,7 @@ class RenderEditable extends RenderBox with RelayoutWhenSystemFontsChangeMixin {
 
     final Offset startOffset = _textPainter.getOffsetForCaret(
       TextPosition(offset: _selection.start, affinity: _selection.affinity),
-      Rect.zero,
+      _caretPrototype,
```
Comment (GaryQian) [inline packages/flutter/lib/src/rendering/editable.dart]: This particular usage does not care what is passed in for the caretPrototype parameter. We pass in the _caretPrototype to not mess up the cached values and avoid redundant re-computation.

File: packages/flutter/lib/src/painting/text_painter.dart [hunk 2]
```diff
@@ -170,6 +170,8 @@ class TextPainter {
   void markNeedsLayout() {
     _paragraph = null;
     _needsLayout = true;
+    _previousCaretPosition = null;
```
Comment (GaryQian) [inline packages/flutter/lib/src/painting/text_painter.dart]: clear cached values when marking as dirty.

**Review by justinmc — COMMENTED** [review]
Looks like a solid performance improvement, good catch.  Can you test it?

**Comment by liyuqian** [comment]
@GaryQian : a performance test for this would be nice (see `*perf*` in https://github.com/flutter/flutter/blob/master/dev/devicelab/manifest.yaml, where `xyz_perf_abc` in that yaml corresponds to a `xyz_perf_abc.dart` in our repo). More details of how to write a performance test is available in [this doc](https://docs.google.com/document/d/1UolaKhEWieP2DDOhFuVQOuR5e545mO2obXcHH6p40Y4/edit#bookmark=id.70t9gnas5d1j).

**Comment by GaryQian** [comment]
Added a new perf test for textfields,

**Review by justinmc — APPROVED** [review]
LGTM

Thanks for doing all of the perf test work!

**Review by liyuqian — APPROVED** [review]
LGTM. Thank you Gary for the performance improvement and the test that guards it! Nit request: can you please put in the commit description the performance number of your test before and after your fix? Just in case that someone regressed it in the future, they'll have some historical records to dig up.

**Comment by liyuqian** [comment]
@GaryQian : the original issue (https://github.com/flutter/flutter/issues/24522) complained about the GPU thread frame rasterization time, but I only see big UI thread frame build time improvement in [your experiment](https://github.com/flutter/flutter/pull/46720#issue-351612188). Do you know why GPU thread frame rasterization time doesn't have a big difference?

**Comment by anuaroa** [comment]
Hi,
The version of installed Flutter is 1.17.1 (Flutter (Channel stable, v1.17.1, on Mac OS X 10.15.4 19E266, locale ru-KZ))
It's been written here https://flutter.dev/docs/development/tools/sdk/release-notes/changelogs/changelog-1.17.0 about #46720 issue, but the problem exists yet.

**Pull request merged**